provider/kubernetes/client.go

```diff
@@ -110,16 +114,53 @@ func createClientFromConfig(c *rest.Config) (*clientImpl, error) {
 	return newClientImpl(clientset), nil
 }
 
-// WatchAll starts namespace-specific controllers for all relevant kinds.
-func (c *clientImpl) WatchAll(namespaces Namespaces, stopCh <-chan struct{}) (<-chan interface{}, error) {
+// WatchNamespaces starts a controller to watch for namespace events.
+func (c *clientImpl) WatchNamespaces(namespaces Namespaces, stopCh <-chan struct{}) (<-chan interface{}, error) {
 	eventCh := make(chan interface{}, 1)
+	eventHandler := c.newResourceEventHandler(eventCh)
+
+	c.namespaceFactory = informers.NewSharedInformerFactory(c.clientset, resyncPeriod)
+	c.namespaceFactory.Core().V1().Namespaces().Informer().AddEventHandler(eventHandler)
 
+	// If no namespaces are specified
 	if len(namespaces) == 0 {
-		namespaces = Namespaces{metav1.NamespaceAll}
-		c.isNamespaceAll = true
+		// If no namespacelabels are specified
+		if c.namespaceLabelSelector.String() == "" {
+			namespaces = Namespaces{metav1.NamespaceAll}
+			c.isNamespaceAll = true
+		} else {
+			// namespacelabels are being used, watch all namespaces for events, use them to get namespaces to watch..
```

Comment:
`ETOOMANYCOMMAS` :) How about

`// namespacelabels are being used: watch all namespaces for events and use them to get namespaces to watch.`